### ISSUE TITLE
Performance Issues with React Native Bridgeless Architecture

### ISSUE THREAD

**Issue by santhoshkumarvgds** [facebook/react-native#47490]
### Description

The **Bridgeless Architecture** in React Native 0.74 and later introduces significant performance issues, particularly for apps with complex logic and high demands for UI responsiveness. Here are the main concerns:

### 1. Screen Load Times
- **Problem**: Synchronous communication with the native runtime slows down screen loading, particularly when multiple API calls are made.
- **Impact**: Initial loads are bottlenecked, reducing overall app responsiveness.

### 2. UI Responsiveness and Touch Lag
- **Problem**: The switch to synchronous communication impacts touch response, causing a laggy and less fluid user experience.
- **Result**: Reduced interaction quality and increased user frustration.

### 3. Lazy Module Loading
- **Default Behavior**: Bridgeless architecture loads modules lazily by default, contributing to delayed responses in complex pages.
- **Partial Fix**: Disabling lazy loading improves response slightly but doesn’t fully solve the underlying delay issues.

## Comparative Performance with Older Versions

- **Testing Results**:
  - **React Native 0.71.8** (asynchronous bridge): Modal with **32 API calls** loads in approximately **3 seconds**.
  - **React Native 0.74.x / 0.76.1** (bridgeless): Modal with only **14 API calls** takes over **10 seconds**.

## Attempts to Improve Performance

- **Native Threads & Configuration Adjustments**: Shifting API calls to native threads, and enabling/disabling options like **Hermes** and **New Architecture**, failed to yield significant performance improvements.

## Call for Optimizations

The **React Native community** and core team are urged to provide optimizations or alternative solutions to enhance performance under the bridgeless architecture. 

**Feedback Needed**: Community support and solutions for the architectural limitations causing these performance issues.

### Steps to reproduce

1. **Create a page/modal** with 10-30 API calls.
2. **Test in React Native 0.74.x** or later (Bridgeless).
3. **Measure load time** and UI responsiveness.
4. **Compare with 0.71.8** (Async Bridge).

Expected: Slower performance in the bridgeless versions.

### React Native Version

0.74+

### Affected Platforms

Runtime - Android, Runtime - iOS

### Areas

Bridgeless - The New Initialization Flow, Other (please specify)

### Output of `npx react-native info`

```text
System:
  OS: macOS 14.5
  CPU: (8) arm64 Apple M1
  Memory: 206.23 MB / 16.00 GB
  Shell:
    version: "5.9"
    path: /bin/zsh
Binaries:
  Node:
    version: 20.14.0
    path: /usr/local/bin/node
  Yarn:
    version: 1.22.19
    path: /usr/local/bin/yarn
  npm:
    version: 10.7.0
    path: /usr/local/bin/npm
  Watchman: Not Found
Managers:
  CocoaPods: Not Found
SDKs:
  iOS SDK: Not Found
  Android SDK: Not Found
IDEs:
  Android Studio: 2021.3 AI-213.7172.25.2113.9123335
  Xcode:
    version: 16.1/16B40
    path: /usr/bin/xcodebuild
Languages:
  Java:
    version: 17.0.11
    path: /usr/bin/javac
  Ruby:
    version: 2.7.4
    path: /Users/santhoshkumarvgds/.rvm/rubies/ruby-2.7.4/bin/ruby
npmPackages:
  "@react-native-community/cli":
    installed: 15.0.0
    wanted: ^15.0.0
  react:
    installed: 18.3.1
    wanted: 18.3.1
  react-native:
    installed: 0.76.1
    wanted: ^0.76.0
  react-native-macos: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: true
  newArchEnabled: false
iOS:
  hermesEnabled: Not found
  newArchEnabled: false
```

### Stacktrace or Logs

```text
The issue occurs when multiple API calls are made on initial screen load. The app becomes unresponsive and takes significantly longer to load the screen. Despite attempts to optimize by disabling lazy loading and using native threads, the issue persists. No crash or stack trace is generated, but the app's performance degrades considerably.
```

### Reproducer

Reproducer: Unfortunately, I cannot provide a public repository or Expo Snack at this time. Please let me know if you would like further details or an isolated example of the issue.

**Comment by arasrezaei** [comment]
I agree, there are lots of lags and crashes.
When I turn newArch off, the app is much smoother

**Comment by JustJoostNL** [comment]
@lovegaoshi did you manage to fix that choppy drawer animation? The new architecture heavily slows down my app, is there any timeline for fixing this?

**Comment by lovegaoshi** [comment]
unfortunately no. im reliant on react navigation and react native paper,
and both have minor but unacceptable UX issues that i could not upgrade to
the new arch with. right now im reinventing the wheel to write my own ui
lib and migrate away from these ui libs.

given other important libs like rn firebase or notifee also breaks, its
peobably the best to not chase the newest feature.

**Comment by YasinAkimura** [comment]
I can also confirm that the new arch is causing a few performance issues regarding responsiveness,
and I think react-navigation / expo-router are not ready for the new arch yet.

I can notice a significant lag in catching touch events when doing things quickly which may indicate the blocking nature of synchronous events?

Also a few bugs where Views seem to always fade / animate when layout changes or Text with adjustToFitSize and "unacceptable" lineHeight completely bug out but these are different issues.

**Comment by thespacemanatee** [comment]
I just wanted to add that on our project (which is quite massive but well-maintained), we also see significant performance degradations in debug mode on RN 0.76, even with the new architecture turned **_off_**. The problem might be related to hermes, which has steadily declined in performance over the past few releases. Can anyone confirm that 0.76 + new arch disabled also causes performance regressions?

**Comment by lovegaoshi** [comment]
okay, im probably able to provide a repo soon(tm). i have a MotionLayout
equivalent component just built from scratch using reanimated and gesture
handler, and new arch is noticably choppier than old.

**Comment by lovegaoshi** [comment]
[here](https://github.com/lovegaoshi/rnbug-newarch-perf). I have a MotionLayout (black bar at the bottom) that drags to expand to fullscreen. This is fluid with the old arch, new arch without expo's stack component, new arch with flatlist, but choppy with new arch **and** expo's stack **or** flashlist.


im not sure how useful this actually is, as this is an expo template (nowadays loaded with expo bloats. i used it for convenience) rather than a reproducer, and technically all react native components work as expected, just 3rd party libs are not. 

not sure who to ping to get visibility.

**Comment by C0rren** [comment]
We have the same issues in our application. Notably however, our Android app runs much smoother than iOS.

**Comment by katche70** [comment]
Unfortunately, I have to completely agree with this. I spent the whole week, full of anticipation and euphoria, painstakingly converting our app to the new architecture. 
I am now completely disillusioned. There are a lot of places where you notice it's really stuck and you just can't get to the previous version on iOS. A lot of marketing for nothing. What a shame ☹️

**Comment by thespacemanatee** [comment]
@cortinico @cipolleschi any chance the React Native team can look into this with the reproducer from @lovegaoshi? It's not feasible for me to provide my company repo for obvious reasons, and creating a new project is not going to truly reflect the performance regressions in more complex applications. I find it hard to believe that Meta is using 0.76 internally unless it is really a case of user error, but it seems unlikely as many people are facing this issue.

To reiterate, here is my "minimally reproducing setup" on a complex application:

- RN 0.76
- Expo SDK 52
- Tried both New Architecture ON and OFF but performance impact is identical
- Tried both `hermes-2024-11-12-RNv0.76.2-5b4aa20c719830dcf5684832b89a6edb95ac3d64` and `hermes-2024-08-15-RNv0.75.1-4b3bf912cc0f705b51b71ce1a5b8bd79b93a451b` but performance impact is identical

Performance is terrible on debug builds but seemingly faster than 0.75 on production, so the performance impact is only felt in debug mode.

I believe 0.76 has been the most insidious upgrade cycle because the changes according to the [upgrade helper](https://react-native-community.github.io/upgrade-helper/?from=0.75.4&to=0.76.2) are minimal, but there are so many invisible moving parts under-the-hood that makes it impossible for us to have a reasonable timeline for any action items to pinpoint any sort of root cause. Many of us have to build features for our companies and upgrading should not be this tedious.

Although I greatly appreciate the work that the Meta team have put into this release, it has been an impossible one for us, and this is very worrying for the future of my company project due to the tight deprecation cycles in the React Native ecosystem.

**Comment by thespacemanatee** [comment]
@cipolleschi Thank you so much for the response.

> Is this happening for Android, iOS, or both of them?

Both of them are slow in debug mode. Both are fast in production mode.

On our end, we will try to reverse engineer and determine whether an external dependency is causing this debug performance regression. The current list of dependencies we are using is reputable though as far as I can tell, comparing it against https://reactnative.directory/.

Also, CMIIW, but Bluesky is [still using RN 0.74](https://github.com/bluesky-social/social-app/blob/f1c221ed2848d156d81816fbf9aa5d5580a6aacd/package.json#L170)? Do they have a branch where they have validated that 0.76 + New Arch runs smoothly?

**Comment by C0rren** [comment]
Small consolation, but for those of you running react navigation as your navigation package, it appears that native stack navigators run faster than "normal" stack navigators for screen transitions.

**Comment by cipolleschi** [comment]
@thespacemanatee can you try something for me. On iOS, can you try to modify the [`node_modules/react-native/scripts/cocoapods/utils.rb` file](https://github.com/facebook/react-native/blob/0.76-stable/packages/react-native/scripts/cocoapods/utils.rb#L665-L668) like this:

```diff
    def self.add_ndebug_flag_to_pods_in_release(installer)
        ndebug_flag = " -DNDEBUG"

        installer.aggregate_targets.each do |aggregate_target|
            aggregate_target.xcconfigs.each do |config_name, config_file|
-                is_release = config_name.downcase.include?("release") || config_name.downcase.include?("production")
-                unless is_release
-                    next
-                end
                self.add_flag_to_map_with_inheritance(config_file.attributes, 'OTHER_CPLUSPLUSFLAGS', ndebug_flag);
                self.add_flag_to_map_with_inheritance(config_file.attributes, 'OTHER_CFLAGS', ndebug_flag);

                xcconfig_path = aggregate_target.xcconfig_path(config_name)
                config_file.save_as(xcconfig_path)
            end
        end

        installer.target_installation_results.pod_target_installation_results.each do |pod_name, target_installation_result|
            target_installation_result.native_target.build_configurations.each do |config|
-                is_release = config.name.downcase.include?("release") || config.name.downcase.include?("production")
-                unless is_release
-                    next
-                end
                self.add_flag_to_map_with_inheritance(config.build_settings, 'OTHER_CPLUSPLUSFLAGS', ndebug_flag);
                self.add_flag_to_map_with_inheritance(config.build_settings, 'OTHER_CFLAGS', ndebug_flag);
            end
        end
    end
```

Then reinstall pods and try again?

What this does is basically to set the NDEBUG flag to every build config. This should disable some expensive runtime checks we are doing in debug mode to ensure the correctness of the framework. Those checks are disabled in production so that could be one of the causes why Debug is slower than development.

---

AFAIK, Bluesky made an [Android release](https://github.com/bluesky-social/social-app/releases/tag/1.92.0-na-rc.2) with the New Architecture that they asked users to try. I _think_ they are also doing a phased rollout using the Play Store, but I am not completely sure.

**Comment by thespacemanatee** [comment]
@cipolleschi Thanks for the suggestion but unfortunately it's still as slow as before, though at least we can probably rule out NDEBUG as the root cause.

**Comment by cipolleschi** [comment]
To set expectation, I'll be off next week and @cortinico will be back on Thursday.

We will try to prioritize this once we are back.

**Comment by brentvatne** [comment]
@thespacemanatee - i've been speaking regularly with the bluesky team, they intended to launch it recently but they're currently dealing with the issue of having 1 million daily new users - which is higher priority for them 😂  they've released a new arch version to beta testers on github and it has gone well.

i looked into @lovegaoshi's issue and i don't believe that there is something there that is particularly actionable on react-native's side yet - this hasn't been isolated to react-native itself, we need to rule out whether this is an issue with reanimated or not.

**Comment by thespacemanatee** [comment]
@brentvatne To be clear, RN 0.76 is fast in production. It's the debug experience that is very slow for us. It's also slow with the new arch disabled. Just trying to narrow it down here so we don't get off track due to the many different possible configurations. We're still looking for the erroneous dependency (if any) in our project.

**Comment by brentvatne** [comment]
@thespacemanatee - i see. it seems like that is worth its own issue - it's difficult to keep track when a single issue about a broad topic like "performance" has several subplots introduced by tangential experiences :)

**Comment by stanislavkovalov** [comment]
I also migrated a project I'm working on to the new architecture (0.76.1) and am having performance issues. Everything works terribly. Almost all the animations that worked smoothly on 0.72.5 work with delays on 0.76.1. Switching from one screen to another also takes time and it happens with delays

**Comment by LukasMod** [comment]
We had to switch new-arch off on SDK 52. It started to be super laggy on low-end androids with release version when new-arch was on. It's simple app e-commerce app with some forms and without a lot of new libs (mosty expo, project strted from sdk 50). We failed to narrow down the problem.

**Comment by hvq1919** [comment]
+1

**Comment by brentvatne** [comment]
if anyone can help to provide minimal reproducible examples of specific perf issues it'd be much appreciated! we'll be able to respond quickly and help with them if so.

**Comment by lovegaoshi** [comment]
welp. I found the answer to my problem. its bc of how the new arch renders. maybe this will help someone out with animation perf issues

My issue was a MotionLayout like view (like a miniplayer) becomes extremely laggy. I have these 2 debugger profiles for benchmarks: oldarch is [~30ms per transition initiation](https://github.com/user-attachments/files/18087371/APM.old.json), [newarch can be 100ms+](https://github.com/user-attachments/files/18087372/APM.new.json). It just so happens my motionlayout is below a react-navigation drawer content with is a **flex container**. the motionlayout used to not interact with it under the old arch; now it does. The drawer content must be constantly rerendering that the heavier the content is, the slower it becomes. the solution is [quite simple](https://github.com/lovegaoshi/azusa-player-mobile/pull/688/commits/e89a2742875a120fc1d81df4d64ca937f508e4fe), fix the view container's height.

I still think there are newarch rendering getchas but im ready to call this a day. the amount of libs breaking with the new arch (eg. [react-navigation](https://github.com/react-navigation/react-navigation/issues/12324) and [react-native-paper](https://github.com/callstack/react-native-paper/issues/4454)) is insane.

**Comment by redfisher26** [comment]
I am also facing similar issues as described above after moving to RN 76.4.
Most notably with libraries such as react-navigation (drawer specifically), react-native-paper, and other issues related to Expo SDK52.

**Comment by ustuncem** [comment]
Tried this on a brand new project with expo SDK 52 and React Navigation v7. Even just changing a screen reduces the 120 FPS by half and it stays like that for the entire stack. Add a TextInput to the mix and app nearly crashes. Will try to reproduce a minimal repo for this tomorrow

**Comment by ustuncem** [comment]
Old Arch (the performant one): https://github.com/ustuncem/performance-v0.76-oldarch
New Arch(the unperformant one): https://github.com/ustuncem/performance-v0.76-newarch

Everything works perfectly whenI disable the old arch. Problem specifically occurs on Android platform. iOS seems to work just fine. I cant understant which library causes this to happen. Also, release apk size is 103MB on the newArch and 75mb on the oldArch.

**Comment by JustJoostNL** [comment]
It's interesting to see that some people only experience this on Android, and others on both Android an iOS. For me the problem happens on both platforms.

**Comment by hosseinmd** [comment]
> Old Arch (the performant one): https://github.com/ustuncem/performance-v0.76-oldarch New Arch(the unperformant one): https://github.com/ustuncem/performance-v0.76-newarch
> 
> Everything works perfectly when I disable the new arch. Problem specifically occurs on the Android platform. iOS seems to work just fine. I can't understant which library causes this to happen. Also, release apk size is 103MB on the newArch and 75mb on the oldArch.

Is there any review process on going for these reproduced repos? I tested these repos, issue is valid. App freezing after a few seconds, work with it.
I don't think that is because of source code. There is nothing special. Maybe it is because of a package.

**Comment by mmttt89** [comment]
> Old Arch (the performant one): https://github.com/ustuncem/performance-v0.76-oldarch New Arch(the unperformant one): https://github.com/ustuncem/performance-v0.76-newarch
> 
> Everything works perfectly when I disable the new arch. Problem specifically occurs on the Android platform. iOS seems to work just fine. I can't understant which library causes this to happen. Also, release apk size is 103MB on the newArch and 75mb on the oldArch.


Any updates on this? I can confirm, I got the crash on the new arch of this repo.

**Comment by cipolleschi** [comment]
Please, don't ask for updates, as it does not help the investigation in any way.

---

Thanks @ustuncem for the reproducers and @hosseinmd for verifying that they are valid. We will work on them.
I believe you all can understand that this specific period of the year things are moving more slowly than usual, with several people that are off to spend time with their family. 
We appreciate and thank you all for your patience, we will try to do our best to find a solution for it.

I can see that the app is pretty complex, with several modules like:
- Home
- Onboarding
- Authorization
- Authentication

Is there a way to trigger the issue reducing the app surface? The problem might likely live in a specific library. It would be optimal if we can reduce the reproducer to something very simple where fewer libraries are involved, perhaps just Reanimated, Screen and Gesture Handler. If possible, we should remove Expo Router and other layers of abstraction.
As you can imagine, the more components are involved, the harder it is for us to find the actual root cause of the problem.

**Comment by ustuncem** [comment]
Hello @cipolleschi, I will try to reduce the number of packages, but right now only the onboarding and authentication features have some components in them; the rest are empty. I will try to remove the Expo Router over the weekend, but the issue especially arises when navigating to other screens. iOS looks fine to me, but on Android, after a few clicks on any screen, the app freezes and crashes with no logs. When I use adb logcat, it prints memory leaks. My guess is that the navigation and bottom sheet libraries I use are not fully compatible with the new architecture.

Anyway, I will try to reduce the code and make it simpler over the weekend

**Comment by cipolleschi** [comment]
Thanks for spending time on that. From what you are explaining, it looks like that either RNScreens or React Navigation are creating the memory leaks. Both are used by Expo Router under the hoods.
I'll try to set up those two first, to see if they are enough to repro the issue.

**Comment by hosseinmd** [comment]
Can it be inferred that the issue is of this kind?
https://github.com/facebook/react-native/pull/47634
I should tell, I added a patch of this fix to my own project which has the same issue but it not fixed.

**Comment by ustuncem** [comment]
@hosseinmd I dont think it is the problem as the demo only consists of two screens and the issue is happening before the transition. Right there on the first render.

**Comment by ustuncem** [comment]
@cipolleschi Hello, another performance issue that I noticed: On Android, if the user somehow is able to come to the login or register screens without any crushes and enters the information necessary, a bottom sheet pops up. I am using the @gorhom/bottom-sheet package and the time it finishes animating, it almost instantly crushes 95% of the time. That is a library that highly depends on Gesture Handler and Reanimated. You can test it by skipping the onboarding screen and then on the register (with the title "Kayıt Ol"), click the link on the bottom to get to the login screen. Then enter a random phone number and wait for the bottom sheet to slide up. If a crush doesn't occur, performance is highly degraded, FPS is on the 15-20 rates. Both on the UI and JS threads. Maybe it has something to do with Reanimated and Gesture Handler also? Should I also open issues on that repos as well?

**Comment by hosseinmd** [comment]
@ustuncem I am not sure, but it seems problem is because of keyboard controller and bottom sheet. Did you find anything?

@cipolleschi I need to know why this is happening because in my own project I am not using keyboard controller. Perhaps there is something similar we could fix it.

**Comment by JustJoostNL** [comment]
I don't think it's necessarily the animations. Just when scrolling with a basic `ScrollView` the FPS already drops to like 45. But animations are very slow too, and make the JS thread FPS drop to as low as 10. 

Did anyone find a solution to get back to normal performance in development builds on iOS?

**Comment by pleerock** [comment]
I can also confirm updating from 0.75 (old arch) to the latest 0.76 (new arch) caused huge visible performance degradation (checked on iOS). Project is huge (>100k LOC) and obviously I cannot share it. However I'm able to make a call session with RN team and screen-share the results before & after.

**Comment by ustuncem** [comment]
Hello @hosseinmd I am swamped with work so I can't remove the keyboard controller and try the release without it. Is the repro performant after removing the keyboard controller? Did you try it?

**Comment by hosseinmd** [comment]
I update above code.

**Comment by ustuncem** [comment]
I will test it but from what you said issue relies with what keyboard controller is doing internally

**Comment by itsnyx** [comment]
same issue here upgraded from 0.74.5 to 0.76.6 i only been able to test android both production and debug is laggy spacilly on large lists that i use flashlist from shiopify when scrolling fast it shows alot of blank screen 
also one thing that i noticed in debug mode app uses nearly 1 gb of ram! in android which for my complex application was 500 mgb in 0.74.5

**Comment by saidamir101** [comment]
It's not about keyboard controller or any list implementation. There are definitely problems in a new architecture. Having the same issues in our project after upgrade to 0.76.5. App become extremely laggy, this is especially visible with reanimated animations. When new architecture is disabled on 0.76 everything works fast as before.

**Comment by ustuncem** [comment]
What I am curious about is whether any other projects have experienced this during migration. Can anyone with a complex app confirm that they have migrated successfully without lags or performance degradation?

**Comment by hosseinmd** [comment]
Perhaps react-native members think this is other packages responsibility because  their bluesky app working correctly, maybe this should be clarified first of all.

**Comment by cortinico** [comment]
> What I am curious about is whether any other projects have experienced this during migration. Can anyone with a complex app confirm that they have migrated successfully without lags or performance degradation?

For context: Facebook app has been running with the New Architecture since years and we haven't faced any performance regression. Similarly, several other apps from Meta listed here https://reactnative.dev/showcase are running New Architecture/Bridgeless mode and again, we haven't noticed any performance regression.

**Comment by ustuncem** [comment]
Hello @cortinico thanks for the response. Are we doing something that we aren't supposed to then? Are these apps using packages like reanimated, gesture hander etc. from the example repo?

**Comment by talaikis** [comment]
> What I am curious about is whether any other projects have experienced this during migration. 

Each RN version since 0.72 is worse than version before in terms of performance mainly due to JS thread overload. Moving parts of heavy JS code to native improves things, but that's all with old arch. New arch never worked, a lot of lags, constant freezes, etc. Bridgeless crashes immediately after start. An app has a of libs, mostly from expo.

**Comment by lovegaoshi** [comment]
im unfortunately plagued by another performance issue (reanimated when using RN-video as a background element) so im back to this grind. but Ive resolved a performance issue b4, where react-native-navigation's drawer will **recalculate its height every time even when hidden if you have an expandable element that resizes it, causing a performance degradation**. in other words I have a miniplayer that expands, and a hidden drawer behind it; the drawer will cause severe miniplayer expand/collapse lag. [fix is set the height in its style to be a constant number.](https://github.com/lovegaoshi/azusa-player-mobile/pull/688/files)

its a difficult problem that roots from meta, but ultimately caused by 3rd party libs that meta has no control over (or they elected to be, after meta spun out their libs to be 3rd party supported). unfortunately we are at this stage of pointing fingers where meta strictly only supports whatever their core lib is (which is actually fine by all means), 3rd party is often unmaintained (look at react-native-paper's new arch support fiasco, and many others that arent even doing it yet) and probably is fine until you and me the devs combine them in a weird manner, and all we saw is simply switching the newarch flag caused a huge performance degradation. 

lastly there is probably nothing anyone can do except for you and me the devs. identify what element is lagging, literally remove everything else and decide if its still lagging; if so make it into a reproducer here. else slowly add the elements back. this is debug 101 that we all should be familiar of at this point. get the chrome profiler and start profiling.

**Comment by cortinico** [comment]
> Are we doing something that we aren't supposed to then?

Sadly that's kinda hard to say.

I'm reading this conversation daily and I'm seeing generic messages about 'performance regressions' between old and new arch, without metrics or actual proof that there is actually a regression.

Historically, we always had folks in the community sharing that newer version 'feels slower' than older one. However, we validated with our partners and with real metrics that this is not the case.

During 2023/2024 we fixed a number of performance regressions that were reported by the community, but those came with a isolated reproducer + with graphs showing how the framerate dropped before/after.

Sadly, just commenting on this issue saying that the app feels 'slower' is not doing any help. The best we can all do is isolate a reproducer and identify a small set of libraries or APIs that are regressing (if any) on the New Architecture.

**Comment by cipolleschi** [comment]
For performance in debug mode, we had a [lengthy investigation](https://github.com/bluesky-social/social-app/pull/6755#issuecomment-2577555675) on the Bluesky app that seems to point to the `<Profiler>` component.

If you are using this component in your app, can you try to remove it and report back whether it improves the performance? Meanwhile, we are investigating why this is slow and how can we improve it. 

This could improve the performance when the app is in Debug mode, not in Release mode.

**Comment by hosseinmd** [comment]
We are not using the Profiler.
I sent a reproduced code above. Why you won't check it?

**Comment by rubennorte** [comment]
> React Native 0.71.8 (asynchronous bridge): Modal with 32 API calls loads in approximately 3 seconds.
> React Native 0.74.x / 0.76.1 (bridgeless): Modal with only 14 API calls takes over 10 seconds.

Just to clarify, what do you mean with N API calls? Are those network calls? Is the performance issue still reproducing for you if you don't make network requests?

(Btw, if they're network requests, dispatching 32 network requests in any client interaction seems like a bad pattern that will lead to a bad experience in general).

**Comment by kirillzyusko** [comment]
>  I created a new react-native project with new arch and added your login page. it has performance issue yet and app freeze after few seconds, but after i remove keyboard controller everything work perfectly.
i don't know why keyboard controller has this issue. @cipolleschi Could you check this package?

@hosseinmd @cipolleschi I'll have a look on that 👀

**Comment by kirillzyusko** [comment]
@hosseinmd okay, I took a project @ustuncem and pasted your code (keeping `KeyboardProvider` in place) - everything works well:

https://github.com/user-attachments/assets/de101b4d-c490-48f0-aaf3-e12c96c0fe53

@hosseinmd may I kindly ask you to upload your project to GitHub and share a video what exactly you are doing to get a crash? Using your code I'm not able able to reproduce any freezes 🤷‍♂ 

Alternatively you can go to `KeyboardProvider` and remove lines one-by-one. `KeyboardProvider` simply receives keyboard/TextInput events and stores them as Animated/Reanimated values. Maybe frequent updating of these values trigger a crash?

**Comment by kirillzyusko** [comment]
@hosseinmd Okay, it seems like some styles were missing - if I add `style={{ width: 200, height: 50, backgroundColor: 'yellow', marginBottom: 10 }}` to `BottomSheetTextInput` then I can receive an application freeze.

After quick debugging it seems like it happens only if Reanimated handler is attached to a view (i. e. `onKeyboardMoveReanimated={keyboardHandler}` in `KeyboardProvider`).

I think it's reanimated issue, but I will try to debug it further tomorrow to have a 100% confidence in my words 🙂

**Comment by ustuncem** [comment]
@kirillzyusko thank you so much for looking into this.

**Comment by bartlomiejbloniarz** [comment]
@kirillzyusko This could be a recursive commit problem. There is a fix for that in the newest RN version.

**Comment by hosseinmd** [comment]
@bartlomiejbloniarz could you add the link of fix please

**Comment by bartlomiejbloniarz** [comment]
I think [this PR](https://github.com/facebook/react-native/pull/47665) might help, but I didn't check.

**Comment by kirillzyusko** [comment]
@bartlomiejbloniarz I confirm, if I update `react-native` to `0.76.6` then the issue is gone and app not freezes/crashes anymore.

@hosseinmd @ustuncem can you try to update RN to `0.76.6` and see if the issue is still present in reproduction and your projects?

**Comment by kirillzyusko** [comment]
@JustJoostNL the repro provided by @ustuncem contained information about Android issues only. I tested iOS and it works well there. Do you have a reproduction example which shows the problem on iOS?

**Comment by ustuncem** [comment]
@kirillzyusko @bartlomiejbloniarz  hello, I will test it as soon as I can thanks for the update. I am not using any drawer navigation so maybe that's why my repro didn't experience any performance issues on iOS. As @lovegaoshi pointed out here:

> ~im unfortunately plagued by another performance issue (reanimated when using RN-video as a background element) so im back to this grind. but Ive resolved a performance issue b4, where react-native-navigation's drawer will **recalculate its height every time even when hidden if you have an expandable element that resizes it, causing a performance degradation**. in other words I have a miniplayer that expands, and a hidden drawer behind it; the drawer will cause severe miniplayer expand/collapse lag. [fix is set the height in its style to be a constant number.](https://github.com/lovegaoshi/azusa-player-mobile/pull/688/files)~
> 
> its a difficult problem that roots from meta, but ultimately caused by 3rd party libs that meta has no control over (or they elected to be, after meta spun out their libs to be 3rd party supported). unfortunately we are at this stage of pointing fingers where meta strictly only supports whatever their core lib is (which is actually fine by all means), 3rd party is often unmaintained (look at react-native-paper's new arch support fiasco, and many others that arent even doing it yet) and probably is fine until you and me the devs combine them in a weird manner, and all we saw is simply switching the newarch flag caused a huge performance degradation.
> 
> lastly there is probably nothing anyone can do except for you and me the devs. identify what element is lagging, literally remove everything else and decide if its still lagging; if so make it into a reproducer here. else slowly add the elements back. this is debug 101 that we all should be familiar of at this point. get the chrome profiler and start profiling.
> 
> ===== just confirmed I have yet another issue with expanding/collapsing a miniplayer with a view in the background with flex:1. [fix is simple](https://github.com/lovegaoshi/azusa-player-mobile/pull/724). throwing it out here in case it helps anyone. only happens to the new arch.
> 
>  Screencast.from.01-14-2025.11.09.33.AM.webm

It maybe because of how views with uninitted heights interact with the parent content that is provided from @react-navigation/drawer. Another project that I am currently developing is migrated to react-native 0.76 without enabling the new arch. I also updated the react-navigation packages to version 7.x.x. I immediatly noticed performance issues even when not enabling the new arch. Drawer's opening and closing animations are extremely laggy on debug(Haven't tested in production yet.) but when downgrading drawer navigation to version  6.7.2, lags are gone. So this maybe about what changed in @react-navigation/drawer on version 7.x.x rather than a new arch problem. Should we open an issue there as well?

**Comment by ustuncem** [comment]
Tested on iOS also:
With React Native 0.76.6, the new architecture enabled, and @react-navigation/drawer version 7.1.1:

https://github.com/user-attachments/assets/6121ea17-f97f-4740-b357-693bf519e7f2

With React Native 0.76.6, the new architecture disabled, and @react-navigation/drawer version 7.1.1:
https://github.com/user-attachments/assets/1a49e6a4-426c-4ec7-a0ac-65758a3f51b7

Disabling the new architecture significantly reduces flickering. I will try to provide a reproduction example for this issue. Unfortunately, I can't open source the project as my company doesn't allow it.

I'm unsure where to report this issue. Should I continue discussing it here, or should I open an issue in the react-navigation repository? Any guidance would be greatly appreciated.

**Comment by ustuncem** [comment]
I've been testing it for more than 10 minutes and issue seems to have been gone. No crashes/freezes happening on Samsung Galaxy S22. Thanks for looking into this. At least it seems like I can ship it in new arch now.

> @bartlomiejbloniarz I confirm, if I update `react-native` to `0.76.6` then the issue is gone and app not freezes/crashes anymore.
> 
> @hosseinmd @ustuncem can you try to update RN to `0.76.6` and see if the issue is still present in reproduction and your projects?

**Comment by ustuncem** [comment]
Sorry for posting back to back but maybe this helps someone solve the issue here and I need to confirm if this is an expected behaviour. Enabling new architecture adds an extra 21 MB to release apks on Android. Is this valid?

Before enabling the new architecture:
![Screenshot_20250121_192515_Settings.jpg](https://github.com/user-attachments/assets/12657152-ce7b-4039-816a-aefa9ba63740)

After enabling the new architecture:
![Screenshot_20250121_192416_Settings.jpg](https://github.com/user-attachments/assets/d6a9fde7-bdf9-467c-b232-c5ef2b55306b)

Before it was 107 MB and after enabling the new arch, apk size is 128 MB. Do we know what does that extra 21MB contain and where is it coming from?

**Comment by Antisane76** [comment]
> Tested on iOS also: With React Native 0.76.6, the new architecture enabled, and @react-navigation/drawer version 7.1.1:
> 
>  Screen.Recording.2025-01-21.at.18.28.53.mov 
> With React Native 0.76.6, the new architecture disabled, and @react-navigation/drawer version 7.1.1: https://github.com/user-attachments/assets/1a49e6a4-426c-4ec7-a0ac-65758a3f51b7
> 
> Disabling the new architecture significantly reduces flickering. I will try to provide a reproduction example for this issue. Unfortunately, I can't open source the project as my company doesn't allow it.
> 
> I'm unsure where to report this issue. Should I continue discussing it here, or should I open an issue in the react-navigation repository? Any guidance would be greatly appreciated.

@ustuncem with new arch enabled does disabling react-native-screens help with the flickering/performance? I noticed the flickering disappeared when using drawer version 6.7.2 on Android with enableScreens(false)

in root of project: 

import { enableScreens } from "react-native-screens"
enableScreens(false)

**Comment by ustuncem** [comment]
> > Tested on iOS also: With React Native 0.76.6, the new architecture enabled, and @react-navigation/drawer version 7.1.1:
> > 
> >  Screen.Recording.2025-01-21.at.18.28.53.mov 
> > With React Native 0.76.6, the new architecture disabled, and @react-navigation/drawer version 7.1.1: https://github.com/user-attachments/assets/1a49e6a4-426c-4ec7-a0ac-65758a3f51b7
> > 
> > Disabling the new architecture significantly reduces flickering. I will try to provide a reproduction example for this issue. Unfortunately, I can't open source the project as my company doesn't allow it.
> > 
> > I'm unsure where to report this issue. Should I continue discussing it here, or should I open an issue in the react-navigation repository? Any guidance would be greatly appreciated.
> 
> @ustuncem with new arch enabled does disabling react-native-screens help with the flickering/performance? I noticed the flickering disappeared when using drawer version 6.7.2 on Android with enableScreens(false)
> 
> in root of project: 
> 
> import { enableScreens } from "react-native-screens"
> enableScreens(false)

Hello, thank you for the response. I will test it tonight and notify you.

**Comment by gaearon** [comment]
>Perhaps react-native members think this is other packages responsibility because their bluesky app working correctly, maybe this should be clarified first of all.

Just to clarify, Bluesky app is not yet working perfectly with the New Architecture. We don't want to migrate until it runs at least as good or better, and for now it's not there yet. I've commented with latest issues we're discovering in https://github.com/bluesky-social/social-app/pull/6755 so this will be an ongoing process until we can confidently cut over.

**Comment by LouisSyfer** [comment]
I'm experiencing significant performance issues on iOS since I upgraded from SDK 51 to 52. Already, I had to disable the new architecture (my application was literally broken). Then, the flashlist lost considerable performance. But still, navigating from one screen to another is sometimes tedious. Anyone else have this problem? Thank you for your feedback.

**Comment by cipolleschi** [comment]
Hi @ustuncem. Regarding
> Before it was 107 MB and after enabling the new arch, apk size is 128 MB. Do we know what does that extra 21MB contain and where is it coming from?

I think that the answer is yes, it is expected (cc. @cortinico who has more Android expertise). The reason is that the New Arch is basically a rewrite of the Old Arch in C++ so we can share code between iOS and Android. The New Arch has some extra code that we need to build and that could account for those extra MB.

--- 

@LouisSyfer have you tried opening an issue on [Expo repository](https://github.com/expo/expo)? 
Flashlist losing performance on the New Arch only or also on the Old Arch?

**Comment by LouisSyfer** [comment]
@cipolleschi hi ! No I did not open an issue on the Expo repository. But I talk about it on Expo Discord server. For the moment, I don't have any answer. And anyway I can't provide an Minimal reproducible example. All I can say : it's was working very well with expo sdk 51. And the loss of performance happens with the old and new arch. Only with my iOS app (on Android it works well!). 
Something tells me this might be from reanimated. This wouldn't be the first time I've had a problem with this package when I upgrade an Expo sdk...

**Comment by mobinni** [comment]
Overall we saw regressions when livestreaming data, `Text` render commit timing on Fabric vs non-Fabric renderer were nearly a 15 FPS difference when streaming at a 1 second interval

**Comment by mobinni** [comment]
Overall from what I have profiled the render speed of Fabric vs Paper is night and day, but I do have some findings which I need to try to create reproductions for:

1. when it comes to RN `Text` updates when live streaming updates from an apollo subscription the render commit seems to be more expensive on iOS, on Android I can't seem to find a big difference 🤔 but in iOS simulator I did see a 15 FPS difference, we had to get around this with `setNativeProps` on a text-input

2. there are also some regressions with RNGH / RN gesture systems conflicting where if you use RNGH and RN combined gestures don't track well

3. Also when it comes to the keyboard avoiding view I notice that when elements dynamically populate, but you select `autoFocus` on an input field, on Android it will not recalculate the layout properly when the keyboard is dismissed

4. Back gestures seem to be better prioritized in Paper vs Fabric (this is most likely a react-navigation specific interaction, but any guidance here would be useful)

**Comment by efstathiosntonas** [comment]
@mobinni can you try react-native-fast-text and check the FPS?

**Comment by pleerock** [comment]
After many investigations I also found issue in HUGE performance degradation after Bridgeless Architecture upgrade related to many `TextView` used in the UI.

Consider a UI component which shows all emojis (unicode character, just a text), every with it's own onPress handler. Now try to display such component several times and you'll see app loading and working very slow (especially animations). Try the same with new arch disabled and you won't have any issues at all.

**Comment by hosseinmd** [comment]
There is a big question. I think members could answer this.
Is there any successful app with new architecture?

**Comment by mobinni** [comment]
I guess Text in particular has issues in Fabric, honestly most of our app was fine in the upgrade, outside of some visual regressions and third party library crashes.

Resource intensive screens (many api calls, streaming updates, etc) seem to have suffered and require different strategies that are GPU based like Skia

It's hard to say if this is RN or a combination of RN and third party libraries

**Comment by wfern** [comment]
Guys, without a repo or snack is IMPOSSIBLE to help. 

@pleerock can't you put that emoji component in a expo snack and share with us? You can create 2, one in SDK 51 (old arch) and another in SDK 52 (new arch), it will have the same code and it's very fast to do.

**Comment by mobinni** [comment]
After digging a bit more I don't actually think Text has regressed itself, I think there's something fishy with react-navigation and screens in the background continuing to update and destroy FPS 🤔 @wfern I will try to put together a reproducer but trying to get something together that emulates a large production app is difficult

**Comment by mobinni** [comment]
@efstathiosntonas It seems to be stable, it doesn't help with the FPS issues I mentioned or the KeyboardAvoidingView not triggering

**Comment by lovegaoshi** [comment]
I've got another pointer, though probably not applicable to many others on this thread.

My app uses a bunch of self-updating hooks while the app is in the background. I found [some of them](https://github.com/doublesymmetry/react-native-track-player/blob/2765d09cb93541d89be94a4809320e4df33c0d47/src/hooks/useProgress.ts#L33) will cause a severe performance degradation when the app is minimized for a long time (5 min) then resumed, only in the new arch. The screen either froze for 5s+ or became a white screen of death. My hypothesis is adding a check if Appstate is not active before the setState will solve this issue.

**Comment by cortinico** [comment]
> Can the React Native maintainers give us updates ? New versions are released, but the same problems persists with performance degradation and fps drops when enabling the new architecture.

I'm following this thread, but I still have to see a reliable reproducer/video that we can look into. 

Without reproducers we can't really help you folks, you might be writing code that this is just not performant. Or there could be a bug inside the core of React Native. Sadly without seeing your projects there is not much we can do.

**Comment by hosseinmd** [comment]
> > Without reproducers we can't really help you folks, you might be writing code that this is just not performant. Or there could be a bug inside the core of React Native. Sadly without seeing your projects there is not much we can do.


Bule sky seems to have same issue too, You could check it.

**Comment by lovegaoshi** [comment]
At least for me, I have a really hard time reproducing my performance issue on a small scale reproducer - it is quite apparent in my app, but when u take out the bells and whistles the performance lag becomes acceptable/unnoticeable. 

With that said if anyone just need a project to show that the new arch has performance issues, heres [mine](https://github.com/lovegaoshi/azusa-player-mobile/tree/dev). I have newarch and oldarch both [compiled on android](https://github.com/lovegaoshi/azusa-player-mobile/releases/tag/dev-build-2025-03-06) with the only difference between them is [simply](https://github.com/lovegaoshi/azusa-player-mobile/blob/1addbef017edbbfe91eb6b23d5c53fe0309a647e/.github/workflows/android-weekly.yml#L204) [enabling the flag](https://github.com/lovegaoshi/azusa-player-mobile/blob/1addbef017edbbfe91eb6b23d5c53fe0309a647e/scripts/newarch.py#L6). Throughout use, the miniplayer animation becomes noticeably laggier when there is media playing in the new arch, on my Samsung S21. All I know as a dev is I switch to newarch with one line, and I see lag. If the old arch works fine I dont really see the point being only me writing crap code. It could be me having a lot of data/assets loaded in the app while I'm using it, it could be my phone is not particularly strong anymore, it could be only via prolonged period of use that eg. the expo people refused to acknowledge my app has issues. 

I do know quite a few newarch workarounds at this point from debugging my mess - one being the flex calculation when there is a resizing event is much slower than old. I have to [pin the view height instead of using flex](https://github.com/lovegaoshi/azusa-player-mobile/blob/dev/src/components/commonui/FlexViewNewArch.tsx) for the new arch. This directly apply to the miniplayer animation I mentioned above and was a major source of performance degradation for my app. Another being [updating states while the app is in the background](https://github.com/doublesymmetry/react-native-track-player/pull/2370#issuecomment-2679671764) will cause a major lag/white screen of death when the app is resumed on android.

**Comment by ustuncem** [comment]
Hello @cortinico,

I previously provided a minimal reproduction but haven't received any feedback on it, except from @kirillzyusko, who suggested trying the new version. After switching to version 0.76.5, the crash issue was resolved. However, performance issues persist on both iOS and Android.

Due to the nature of the projects where we are experiencing these issues, it's quite difficult to create a minimal reproduction. Nevertheless, I have partly open-sourced my project and reduced it as much as possible here: [GitHub Repository](https://github.com/ustuncem/rn-newarch-performance).

This is a New Architecture-enabled project that I have been developing for the past five months, and it relies heavily on @shopify/flash-list and @gorhom/bottom-sheet. While testing, I noticed that screens with only a single <Text> element perform well and work perfectly. However, if you navigate to src/app/(app)/(tabs)/account.tsx, enable ProfileView, and comment out the <Text> component (which I have already done), performance issues immediately appear. Once enabled, navigating between screens becomes extremely laggy.

Digging further, I disabled ProfileView again (which is just a simple ScrollView with some links in it) and navigated to the home screen (src/app/(app)/(tabs)/home.tsx). There, I enabled FilterTagList, which is a FlashList that, as you know, uses C++ at its core. Once enabled, the lagging starts instantly.

Looking even deeper, the @gorhom/bottom-sheet package performs very poorly with the New Architecture. Since it heavily depends on react-native-gesture-handler and react-native-reanimated, it seems to struggle with performance.

In the production-ready version of my app, with all its features enabled, memory leaks occur in the New Architecture, and I see logs related to react-native-reanimated trying to access and modify a shadow node, which causes crashes:
[logs.txt](https://github.com/user-attachments/files/19150817/logs.txt).

I have reduced the base code as much as possible, but in order for you to see the modals and bottom sheets, I have kept the map and the secrets in an .env file. For security reasons, I will change them once you confirm that you have investigated the reproduction.

Any feedback is welcome, and I am happy to make adjustments based on your revisions and suggestions. The issues are similar on Android as well.

Additionally, Dan Abramov commented on this issue, mentioning that Bluesky is also experiencing similar problems and that they will investigate further. Has there been any development on this? Have they identified the root cause of the issue?

**Comment by LouisSyfer** [comment]
The problem may come from react-native-reanimated. I’ve been guessing that for some time. I have had problems with this package in the past, so it would not be really surprising

**Comment by ustuncem** [comment]
> The problem may come from react-native-reanimated. I’ve been guessing that for some time. I have had problems with this package in the past, so it would not be really surprising 

Yeah logs seem to point us in that direction but @lovegaoshi s find is also interesting. Scrollview's that has flex: 1 on their style prop and flexGrow: 1 on their contentContainerStyle seem to cause performance issues because of the way height is calculated on the new arch. Maybe contributers can focus on these two points on the example, I provided both. I kept the ProfileView example especially for that

**Comment by mobinni** [comment]
After a lot of individual profiling of various parts of the ecosystem I came to find most of the performance issues stem from 3rd party libraries @ustuncem we noticed the same behaviour with gorhom, however we are just looking to remove our usage. The library is needlessly complex. 

In terms of navigation becoming laggy I found that this is mostly due to the lack of virtualization (usage of flatlist/sectionlist) for complex screens when profiling with the react profiler. This lag would subside,

**Comment by BenIrving** [comment]
We have specifically seen extreme performance degradation when upgrading from 76 to 77 last week, with new arch enabled. 

Has anyone else noticed degradation specifically around that update?

We have begun to investigate & debug, will report back if we find anything.

**Comment by ustuncem** [comment]
> After a lot of individual profiling of various parts of the ecosystem I came to find most of the performance issues stem from 3rd party libraries [@ustuncem](https://github.com/ustuncem) we noticed the same behaviour with gorhom, however we are just looking to remove our usage. The library is needlessly complex.
> 
> In terms of navigation becoming laggy I found that this is mostly due to the lack of virtualization (usage of flatlist/sectionlist) for complex screens when profiling with the react profiler. This lag would subside,

Hello @mobinni thanks for the response. 

I have been using bottom-sheet since its version 3 was released and it was working fine up until now, new architecture era of the react native. I am fine with migrating and implementing my own bottom sheet solution but I dont think its the root cause of this issue. ShadowNode commitments to update the UI from reanimated is leaking some memory and I think that's the core reason why reanimated based packages aren't working performantly(react-navigation, bottom-sheet, basic libraries that every project installs instantly when any of us inits a react native project). Also I am using FlashList and that's also leaking memory because of the same reason reanimated is not performant. Upon scrolling up and down on FlashList for performance, views that aren't visible on the viewport are removed from the ShadowTree and that's commited also using React Native's ::commit function. Scroll a few times and boom the app crashed and logs that :commit function killed a thread in the system. I don't know which action to take other than provide a minimal reproduction for this issue so if anyone can point me in a direction I am more than happy to take action to solve this problem. If I should talk with the reanimated guys oy flashlist guys or basically anyone that uses the C++ bridge and uses the ShadowNode::commit function, I am happy to start threads there also

**Comment by mobinni** [comment]
@ustuncem Sorry I didn't mean to say that the root cause of this isn't related to the new architecture, personally I have just fought with the bottom sheet library for too long 😅. However it isn't react-native, but reanimated that plays a big role here. 

Which I believe is being addressed here - https://github.com/software-mansion/react-native-reanimated/pull/6776

In terms of the flashlist performance, I have not noticed similar issues, and we use it quite extensively. Happy to take that one offline and see if we can find a fix for your use-case

**Comment by DorianMazur** [comment]
@mobinni did you have the new architecture enabled in versions 0.76 and 0.77?

**Comment by mobinni** [comment]
@DorianMazur yes we did have new arch enabled on both. Most of our performance regressions were honestly either screens doing too much processing that hadn't been maintained and/or bugs with reanimated/screens etc

**Comment by BenIrving** [comment]
> @ustuncem Sorry I didn't mean to say that the root cause of this isn't related to the new architecture, personally I have just fought with the bottom sheet library for too long 😅. However it isn't react-native, but reanimated that plays a big role here. 
> 
> Which I believe is being addressed here - https://github.com/software-mansion/react-native-reanimated/pull/6776
> 
> In terms of the flashlist performance, I have not noticed similar issues, and we use it quite extensively. Happy to take that one offline and see if we can find a fix for your use-case
> 
> @BenIrving We went from 0.76 to 0.77 and did not notice any significant drop-off. Is it isolated to a particular flow in your app or systemic?

It's systemic to all of the apps that we have. Interestingly, upgrading to 78 resolved the issues we saw.

**Comment by DorianMazur** [comment]
@mobinni Do you have any patches for for the mentioned libraries? How did you solve these bugs? It would be really helpful if you could share more details. Thanks!

**Comment by ustuncem** [comment]
Hello @BenIrving are you using expo? I use expo and will try to migrate if the issues are solved in v0.78. Can you share the package versions that you are using so I can try with the working packages, especially reanimated related ones. Thanks in advance

**Comment by ustuncem** [comment]
I think the root cause is mentioned and highly discussed since December here: https://github.com/software-mansion/react-native-reanimated/pull/6776 as @mobinni mentioned. It looks similiar to our problems discussed in this thread also. When the ShadowNode commit function is runned via an external library something happens that causes apps to crash or lag which seems to point us that a lot of libraries that are widely used aren't ready yet for the new architecture and we have to wait for them to get stable and adapt to the new changes. If anyone that has successfully migrated to the new architecture like @mobinni can share their experience and how they solved this pin points it would be amazing for us that are struggling.

**Comment by bartlomiejbloniarz** [comment]
The issue from [6776](https://github.com/software-mansion/react-native-reanimated/pull/6776) isn't really performance related. It's mostly concerned with a wrong state being mounted.

Upgrading to 0.78 might resolve your issues. We had a problem, where Hermes would perform a heavy `StackOverflow` check on Android and that would slow down the app. It was addressed [here](https://github.com/facebook/hermes/issues/1572) and the fix landed in Hermes for RN 0.78. But I'm not sure if that's the issue that has been tormenting everyone here, because it isn't really architecture dependent (and it is present only on Android). So unless you've been comparing architectures across different versions of RN, it shouldn't be the main concern.

**Comment by mobinni** [comment]
@DorianMazur it depends on the package and the use-case, from the bugs we found here are some current guidelines that don't address the root cause but unblock:

- Gorhom bottom sheet: we've just decided to roll our own for now. 
- Appsflyer: I posted a bug resolution [here](https://github.com/AppsFlyerSDK/appsflyer-react-native-plugin/issues/601) 
- RNGH / RN touch systems conflicting: pick one or the other don't use both at the same time
- react-native-collapsible-tab-view: We saw odd performance issues with switching between tabs or crashes, but we have a bespoke configuration wrapping this library so it's hard to say if it's reanimated, the library or something else
- react-native-change-icon: I patched the entry [here](https://github.com/skb1129/react-native-change-icon/pull/115) to support new arch but got no response from the maintainers
- react-native-screens modal height issue: temporary remediation [here](https://github.com/software-mansion/react-native-screens/issues/2587#issuecomment-2668479646) while the team looks into it
- Scrollview touchable area not touchable on low-end android: Wrap it with a view flex 1 (strange issue, not sure if it's RNGH or RN related)
- Crashes on android for flashlist and back events in 0.76 for react-native-screens: update to 0.77 and react-native-screens 4.6+

**Comment by bartlomiejbloniarz** [comment]
@mobinni The issue that I described affects only reanimated (and packages that use reanimated). You would have to build React Native from source and apply the mentioned PR.

**Comment by efstathiosntonas** [comment]
@bartlomiejbloniarz I think there was a line that could be added on `react-native/gradle/libs.versions.toml` to pin the version of the Hermes (so there's not need for applying the PR) but can't recall it, @cipolleschi has mentioned this on another issue iirc.

**Comment by ChinKX** [comment]
My project is also experiencing significant performance degradation after upgrading to **Expo 52 (52.0.29) + RN 76 (0.76.6)** with **new architecture disabled**. However, this only happens in development mode while release mode is working perfectly fine. Our findings is that the [expo dev-plugins are causing this issue](https://github.com/expo/dev-plugins/issues/60). We are not sure of the exact root cause yet but commenting out these dev-plugins seems to solve the performance issue for now. Will update here again if we have any new findings.

**Comment by thespacemanatee** [comment]
@ChinKX I can confirm that removing expo dev-plugins fixed the performance degradations for us as well. This is super huge for us and I would like to sincerely thank you for taking the time to post that here!

**Comment by ChinKX** [comment]
@thespacemanatee Glad that this is working for you! Kudos to my teammates who did the hard work to come up with this solution. For the new arch, we are also not sure if this is related but we are actively looking into this issue since this is just a temporary solution. Will definitely share again if any new findings.

**Comment by brentvatne** [comment]
@thespacemanatee @ChinKX - we've deployed a fix to SDK 52 for this: https://github.com/expo/expo/pull/35581 - this addresses a regression in dev tools plugins with respect to React Native's websockets and transferring binary data (see https://github.com/expo/expo/pull/35581 for more info). you can install the latest version of the expo package and automatically update any of the dev tools plugins that we provide in https://github.com/expo/dev-plugins with `npx expo install --fix`

**Comment by hosseinmd** [comment]
Don't close.

**Comment by evelant** [comment]
I just released my app with the new arch enabled on expo 53 and rn 0.79.2 and now have users complaining of poor performance. I know that's not very helpful information, our app is quite complex and I haven't found potential causes yet, but it's at least another data point that there may still be issues with perf on the new arch.

**Comment by genesiscz** [comment]
Similar for our upgrade from 0.72 -> 0.77 + new arch :(

<img width="785" alt="Image" src="https://github.com/user-attachments/assets/4c95b7be-917f-4188-a605-97b7a6b35373" />

**Comment by mobinni** [comment]
What data-fetching library do you all use? @genesiscz @evelant I have noticed that when doing fetch the timing between resolution and render is pretty bad. A network request that resolves in 200ms takes over 1 second to fully render 🤔 our app is fairly complex and has some network waterfall combined with expensive components to render but even then it seems a bit aggressive. We use apollo / sentry

**Comment by genesiscz** [comment]
> What data-fetching library do you all use? @genesiscz @evelant I have noticed that when doing fetch the timing between resolution and render is pretty bad. A network request that resolves in 200ms takes over 1 second to fully render 🤔 our app is fairly complex and has some network waterfall combined with expensive components to render but even then it seems a bit aggressive. We use apollo / sentry

Redux sagas, rtk query and lately react query - all of them using fetch wrapped with OpenApi generated TS packages. It's a hell :) and we also log with sentry and AppDynamics

**Comment by Rodrigo816** [comment]
> Similar for our upgrade from 0.72 -> 0.77 + new arch :(
> 
> <img alt="Image" width="785" src="https://private-user-images.githubusercontent.com/1512353/451606402-4c95b7be-917f-4188-a605-97b7a6b35373.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDkxMTY1MjksIm5iZiI6MTc0OTExNjIyOSwicGF0aCI6Ii8xNTEyMzUzLzQ1MTYwNjQwMi00Yzk1YjdiZS05MTdmLTQxODgtYTYwNS05N2I3YTZiMzUzNzMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDYwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA2MDVUMDkzNzA5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTIxZTkwYTFkMGI3MWExNDViNDRhMGU1Nzk4Njc2OGY3Y2Q0NmU1MTU3ZjZkNTkxMDhiMDM2M2M2ODNlYzI2YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.SNsfk7GqEh_ukdRXuFuIeoR7STSmATkNIBnAYcHbp40">

How you make that cool metrics?

**Comment by genesiscz** [comment]
> > Similar for our upgrade from 0.72 -> 0.77 + new arch :(

> How you make that cool metrics?

Xcode -> Window -> Organizer

I literally got notification about the Regression from production

![Image](https://github.com/user-attachments/assets/bee05f85-cf24-4256-95eb-77a8f83c4c01)

**Comment by pleerock** [comment]
From my observations, the huge performance difference is visible when you have LOT OF components rendered in the tree. 
For those who have time to create a reproduction repo - just try to create 10k <Text> views (using for-loop) with any text inside (even a single character) in a single scroll list. Then check how it works in old vs new arch. Share the repo and results.

**Comment by evelant** [comment]
Possibly related: 

https://github.com/facebook/react-native/issues/51869
https://github.com/facebook/react-native/issues/50731

**Comment by cortinico** [comment]
> The **Bridgeless Architecture** in React Native 0.74 and later introduces significant performance issues, particularly for apps with complex logic and high demands for UI responsiveness. Here are the main concerns:

We're closing this as this issue is quite old. We invite everyone to try 0.81 which contains several performance improvements especially for New Arch.

If you encounter other performance regression, please open a new issue with a reproducer and we'll be happy to investigate.
